File: 백준/Silver/17086. 아기 상어 2/아기 상어 2.py
import sys
from collections import deque
input = sys.stdin.readline
N, M = map(int, input().split())
ls = [list(map(int, input().split())) for _ in range(N)]
dx = [-1, -1, -1, 0, 0, 1, 1, 1]
dy = [-1, 0, 1, -1, 1, -1, 0, 1]
q = deque()
ans = 0
for i in range(N):
    for j in range(M):
        if ls[i][j] == 1:
            q.append((i, j, 0))

while q:
    x,y,tmp = q.popleft()
    for d in range(8):
        nx = x + dx[d]
        ny = y + dy[d]
        if nx < 0 or nx >= N or ny < 0 or ny >= M:
            continue
        if ls[nx][ny]:
            continue
        ls[nx][ny] = tmp + 1
        ans = max(ans, ls[nx][ny])
        q.append((nx, ny, tmp + 1))
print(ans)

File: 백준/Silver/17086. 아기 상어 2/아기 상어 2.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [N, M] = input[0].split(" ").map(Number);
const ls = input.slice(1).map(line => line.split(" ").map(Number));

const dx = [-1, -1, -1, 0, 0, 1, 1, 1];
const dy = [-1, 0, 1, -1, 1, -1, 0, 1];

let q = [];
let ans = 0;

for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
        if (ls[i][j] === 1) {
            q.push([i, j, 0]);
        }
    }
}

while (q.length) {
    const [x, y, tmp] = q.shift();

    for (let d = 0; d < 8; d++) {
        const nx = x + dx[d];
        const ny = y + dy[d];

        if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
        if (ls[nx][ny] > 0) continue;
        ls[nx][ny] = tmp + 1;
        ans = Math.max(ans, ls[nx][ny]);
        q.push([nx, ny, tmp + 1]);

    }
}
console.log(ans);
